### PR TITLE
Meme gallery

### LINKED ISSUE #5: 5 - Gallery
As a user, I should be directed to a public gallery of the most recent public memes uploaded to the site. This public gallery acts as the home page, and each meme has a caption which includes the uploader and total views.

### PR DESCRIPTION
**Pull request by jphilly11**
A user should be able to see a page of created memes if they go to the landing page which I changed to be public.

### CODE REVIEW

File: api/src/main/java/com/exelaration/abstractmemery/controllers/MemeController.java [hunk 1]
```diff
@@ -21,4 +24,11 @@ public Meme uploadMeme(@RequestBody Meme meme) throws Exception {
 
     return memeService.save(meme);
   }
+
+  @GetMapping("/memes")
```
Comment (mchughmk) [inline api/src/main/java/com/exelaration/abstractmemery/controllers/MemeController.java]: The URL specification here is unnecessary.  Used properly, REST specifies that a `GET` call with no parameters is a "get all" request.  Thus, the URL for this request should be `/meme/`, but this line makes it `/meme/memes/`.

File: api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeMetadataServiceImpl.java [hunk 2]
```diff
@@ -27,4 +33,17 @@ public Meme save(Meme meme) {
   public void deleteById(Integer id) {
     memeRepository.deleteById(id);
   }
+
+  public ArrayList<String> getMemes() {
+    ArrayList<String> memeNames = memeRepository.getMemes();
+    ArrayList<String> memeURLs = new ArrayList<String>();
+    String directory = "/app/src/main/resources/memes/";
```
Comment (KatherynRizek) [inline api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeMetadataServiceImpl.java]: @mchughmk  I know that this is what you recommended this morning, but I think I'm still a little confused on why the file encoding retrieval is in MemeMetadataService. I thought this service was solely to retrieve database information. If so, wouldn't having it grab the encoded file violate the single responsibility part of the service?

Comment (jphilly11) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeMetadataServiceImpl.java]: Should the memeService be the one that makes all the calls to the other services and sends off the list of memeURLs?

File: api/src/test/java/com/exelaration/abstractmemery/services/MemeServiceTest.java [hunk 3]
```diff
@@ -31,7 +32,7 @@ public void init() {
   }
 
   @Test
-  public void save_WhenMemeSavedSuccessfully_ExpectMemeReturned() {
+  public void memeServicesave_WhenMemeSavedSuccessfully_ExpectMemeReturned() {
```
Comment (KatherynRizek) [inline api/src/test/java/com/exelaration/abstractmemery/services/MemeServiceTest.java]: do we need to specify `memeServiceSave` instead of just `save`, the method name, in the test name when it is already in the `MemeServiceTest`?

Comment (spanjshiri) [reply api/src/test/java/com/exelaration/abstractmemery/services/MemeServiceTest.java]: no, although that's how all the tests are named so I just kept the same format

File: api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java [hunk 4]
```diff
@@ -26,4 +27,24 @@ public Meme save(Meme meme) {
...
+  public ArrayList<String> getMemes() {
+    ArrayList<String> memeNames = memeMetadataService.getMemes();
+    if (memeNames == null) {
+      return null;
+    }
+    ArrayList<String> memeURLs = new ArrayList<String>();
+    String directory = "/app/src/main/resources/memes/";
+    for (String memeName : memeNames) {
+      try {
+        String meme = fileStorageService.getFileURL(memeName, directory);
```
Comment (mchughmk) [inline api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java]: Passing in the directory isn't necessary here.  The `fileStorageService` should know where to find the images, we just need to tell it what image to give us.

File: api/src/test/java/com/exelaration/abstractmemery/services/MemeServiceTest.java (hunk 3 above)
Comment (mchughmk) [reply api/src/test/java/com/exelaration/abstractmemery/services/MemeServiceTest.java]: I agree with Katheryn

**Review by mchughmk — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java [hunk 5]
```diff
@@ -22,7 +22,7 @@ public void save(String fileName, String urlData, String storageLocation) {
     try {
       fileLocation = (uploadingDir + fileName);
       Path path = Paths.get(fileLocation);
-      byte[] decodedBytes = Base64.getDecoder().decode(urlData);
+      byte[] decodedBytes = Base64.decodeBase64(urlData);
```
Comment (KatherynRizek) [inline api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java]: Why was this changed to use the Tomcat decoder?

File: api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java [hunk 6]
```diff
@@ -33,4 +33,23 @@ public void save(String fileName, String urlData, String storageLocation) {
...
+  public String getFileURL(String fileName) {
+    String fileURL = "";
+    String storageLocation = "/app/src/main/resources/memes/";
+    File file = new File(storageLocation + fileName);
+    if (file.exists()) {
+      Path filePath = Paths.get(storageLocation + fileName);
+      byte[] fileBytes;
+      try {
+        fileBytes = Files.readAllBytes(filePath);
+        fileURL = Base64.encodeBase64String(fileBytes);
```
Comment (KatherynRizek) [inline api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java]: The encoding can be done with the Java Base64. An example is in the direct links PR

File: api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java (hunk 5 above)
Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java]: idk, what dependency was it using before

File: api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java (hunk 6 above)
Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java]: ok

File: api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java [hunk 7]
```diff
@@ -65,4 +68,21 @@ public void MemeControllerTest_WhenNullMemePosts_ExpectStatus400() throws Except
                 .content(new ObjectMapper().writeValueAsString(testMeme)))
         .andExpect(badRequest);
   }
+
+  @Test
+  public void MemeControllerTest_WhenGetMemes_ExpectStatus200andJSONreturn() throws Exception {
```
Comment (KatherynRizek) [inline api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java]: better name would be `getMemesForGallery_WhenMemesExist_ExpectStatus200andJSONreturned()` so that it matches the method name

File: api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java [hunk 8]
```diff
@@ -65,4 +68,21 @@ public void MemeControllerTest_WhenNullMemePosts_ExpectStatus400() throws Except
...
+    when(memeService.getMemes()).thenReturn(expectedMemeData);
+
+    mockMvc
+        .perform(MockMvcRequestBuilders.get(url))
+        .andExpect(ok)
+        .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON_VALUE))
+        .andExpect(jsonPath("$", hasSize(1)))
+        .andExpect(jsonPath("$[0]", is("TestMemeData")));
+  }
 }
```
Comment (KatherynRizek) [inline api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java:136]: There should probably also be a test that checks if the memes don't exist

File: api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java (hunk 7 above)
Comment (spanjshiri) [reply api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java]: 👍

File: api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java [hunk 9]
```diff
@@ -65,4 +68,37 @@ public void MemeControllerTest_WhenNullMemePosts_ExpectStatus400() throws Except
...
+        .perform(MockMvcRequestBuilders.get(url))
+        .andExpect(ok)
+        .andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON_VALUE))
+        .andExpect(jsonPath("$", hasSize(1)))
+        .andExpect(jsonPath("$[0]", is("TestMemeData")));
+  }
+
+  @Test
+  public void getMemesForGallery_WhenMemeIsNull_ExpectStatus400andJSONreturn() throws Exception {
+    ResultMatcher badRequest = MockMvcResultMatchers.status().isBadRequest();
```
Comment (KatherynRizek) [inline api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java]: I'm sorry to keep fixating on this part but now I have a question. Do we actually want it to be a bad request if there aren't any memes to be displayed? What if a user hasn't made any memes yet? Should it actually be a 200 and just be blank like the direct links ticket (when the image isn't found locally)?

File: api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java [hunk 10]
```diff
@@ -33,4 +33,23 @@ public void save(String fileName, String urlData, String storageLocation) {
   private void createDirectory(String uploadDirectory) {
     new File(uploadDirectory).mkdirs();
   }
+
+  public String getFileURL(String fileName) {
+    String fileURL = "";
+    String storageLocation = "/app/src/main/resources/memes/";
```
Comment (mchughmk) [inline api/src/main/java/com/exelaration/abstractmemery/services/implementations/FileStorageServiceImpl.java]: Isn't this value used in another function in this class?  It should be extracted out to a class variable or constant.

File: api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java (hunk 9 above)
Comment (mchughmk) [reply api/src/test/java/com/exelaration/abstractmemery/controllers/MemeControllerTest.java]: I agree with this.  If there's nothing to display, we just show an empty page, or text that says "No memes yet, upload one to get started."  Something like that.

**Review by mchughmk — CHANGES_REQUESTED** [review]

File: api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java [hunk 11]
```diff
@@ -44,4 +45,21 @@ public String getMeme(int id) {
     }
     return memeData;
   }
+
+  public ArrayList<String> getMemes() {
+    ArrayList<String> memeNames = memeMetadataService.getMemes();
+    if (memeNames == null) {
+      throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Images Do Not Exist");
```
Comment (KatherynRizek) [inline api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: I'm still not sure if we should be actually throwing an error here. I believe the way this is set up, if a new user would go to the home page, they would receive a 404. I think we would just want it to be blank but thoughts?

Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: It shows up as a Meme Not Found in the frontend for the specific image. The rest of the images that do exist still show up

Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: Sorry I may have misinterpreted, let me check something

Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: image

Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: This is what shows up if there are no memes in the db

Comment (KatherynRizek) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: I'm getting a 403

Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: that's because the JWTs are not fixed. The antmatchers are restricting you from accessing it because you are not authorized

Comment (KatherynRizek) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: image

Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: You have to change the antmatchers to "/**" in the websecurity file temporarily to see that it works

Comment (spanjshiri) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: Also there is a typo in the gallery_url. It should be "/meme" not "/meme/". I will fix that

Comment (KatherynRizek) [reply api/src/main/java/com/exelaration/abstractmemery/services/implementations/MemeServiceImpl.java:52]: Yep, that works. Nice

**Review by KatherynRizek — DISMISSED** [review]

**Review by mchughmk — APPROVED** [review]

**Review by KatherynRizek — APPROVED** [review]
